File: packages/ui-flutter/lib/src/password_input.dart
import 'package:flutter/material.dart';

import 'input.dart';
import 'theme.dart';

/// Mirrors `packages/ui/src/components/password-input.tsx`: a
/// [KinetixInput] with a show / hide toggle in the trailing slot
/// (`visibility` / `visibility_off` icons).
class KinetixPasswordInput extends StatefulWidget {
  const KinetixPasswordInput({
    super.key,
    this.controller,
    this.onChanged,
    this.placeholder = '',
    this.isError = false,
    this.enabled = true,
  });

  final TextEditingController? controller;
  final ValueChanged<String>? onChanged;
  final String placeholder;
  final bool isError;
  final bool enabled;

  @override
  State<KinetixPasswordInput> createState() => _KinetixPasswordInputState();
}

class _KinetixPasswordInputState extends State<KinetixPasswordInput> {
  bool _visible = false;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return KinetixInput(
      controller: widget.controller,
      onChanged: widget.onChanged,
      placeholder: widget.placeholder,
      isError: widget.isError,
      enabled: widget.enabled,
      obscureText: !_visible,
      trailing: GestureDetector(
        onTap: () => setState(() => _visible = !_visible),
        child: Icon(
          _visible ? Icons.visibility_off : Icons.visibility,
          size: 18,
          color: c.mutedForeground,
        ),
      ),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/PasswordInput.swift
//
// PasswordInput.swift — KinetixPasswordInput.
//
// Mirrors packages/ui/src/components/password-input.tsx: a KinetixInput
// with a show / hide toggle in the trailing slot. Toggle glyph is the
// `eye` / `eye.slash` SF Symbol (the React source's lucide `Eye` /
// `EyeOff`). Flipping visibility swaps SecureField ↔ TextField, which
// drops keyboard focus — an accepted quirk of this pattern.
//

import SwiftUI

public struct KinetixPasswordInput: View {
    @Environment(\.kinetixColors) private var colors

    @Binding private var text: String
    @State private var visible = false
    private let placeholder: String
    private let isError: Bool

    public init(text: Binding<String>, placeholder: String = "", isError: Bool = false) {
        self._text = text
        self.placeholder = placeholder
        self.isError = isError
    }

    public var body: some View {
        KinetixInput(text: $text, placeholder: placeholder, isError: isError, isSecure: !visible) {
            Button {
                visible.toggle()
            } label: {
                Image(systemName: visible ? "eye.slash" : "eye")
                    .font(.kinetixBody)
                    .foregroundStyle(colors.mutedForeground)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(visible ? "Hide password" : "Show password")
        }
    }
}
